### PR TITLE
Consul name resolution in-memory cache

### PR DESCRIPTION
**Pull request by a-elsheikh**
# Description

Added service cache to consul nr component

## Issue reference

#934

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: nameresolution/consul/consul.go
Comment (yaron2) [inline nameresolution/consul/consul.go:79]: You need a RLock() here, as it's not safe to read and write concurrently.

**Review by yaron2 — CHANGES_REQUESTED** [review]

File: nameresolution/consul/consul.go
Comment (a-elsheikh) [reply nameresolution/consul/consul.go:82]: good catch, I had a second look and did a general refactor in this area

**Comment by yaron2** [comment]
@a-elsheikh have you tested this in your environment?

**Comment by a-elsheikh** [comment]
> @a-elsheikh have you tested this in your environment?

In self-hosted mode, yes - it would be best if someone could evaluate the behaviour on k8s

**Comment by pkedy** [comment]
@a-elsheikh After the 1.3.0 release, I will spend some cycles to test this on K8s.

File: nameresolution/consul/consul.go
Comment (artursouza) [inline nameresolution/consul/consul.go:78]: Have you considered using a LRU cache implementation like https://github.com/hashicorp/golang-lru

File: nameresolution/consul/watcher.go
Comment (artursouza) [inline nameresolution/consul/watcher.go:59]: Is this to keep the cache warm?

**Review by artursouza — CHANGES_REQUESTED** [review]

File: nameresolution/consul/watcher.go
Comment (a-elsheikh) [reply nameresolution/consul/watcher.go:59]: correct, this is using blocking queries to keep relevant services warm in cache - see https://www.consul.io/api-docs/features/blocking

File: nameresolution/consul/consul.go
Comment (a-elsheikh) [reply nameresolution/consul/consul.go:78]: sorry, what benefit would that give over the current implementation?

**Comment by yaron2** [comment]
@artursouza can you continue reviewing this please?

**Comment by a-elsheikh** [comment]
@artursouza any updates on this?

**Comment by a-elsheikh** [comment]
@artursouza @yaron2 any eyes on this please, I've just attempted to merge master again but these failures I can't resolve

**Comment by yaron2** [comment]
@a-elsheikh many apologies this has taken so long to review, we'll actively put eyes on this PR now.

Can you please merge with master? /cc @artursouza

**Comment by a-elsheikh** [comment]
> @a-elsheikh many apologies this has taken so long to review, we'll actively put eyes on this PR now.
> 
> Can you please merge with master? /cc @artursouza

no worries, thank you @yaron2

**Comment by a-elsheikh** [comment]
hey @yaron2 @artursouza - any updates on this please. We're keen on utilizing this feature as part of our adoption strategy

File: nameresolution/consul/consul.go
Comment (artursouza) [inline nameresolution/consul/consul.go:106]: Instead of shuffling to pick one, can you simply pick a random position? What is the advantage of shuffling?

File: nameresolution/consul/consul.go
Comment (artursouza) [inline nameresolution/consul/consul.go:172]: What is the different between remove and expire?

File: nameresolution/consul/consul.go
Comment (artursouza) [inline nameresolution/consul/consul.go:83]: Can you use a sync.Map instead?

**Review by artursouza — CHANGES_REQUESTED** [review]

File: nameresolution/consul/consul.go
Comment (a-elsheikh) [reply nameresolution/consul/consul.go:83]: yes will do, thanks

File: nameresolution/consul/consul.go
Comment (a-elsheikh) [reply nameresolution/consul/consul.go:106]: This is based on a fisher-yates shuffle, it's a non-deterministic method of generating a random result. Consul use this same approach in their dns

File: nameresolution/consul/consul.go
Comment (a-elsheikh) [reply nameresolution/consul/consul.go:163]: `remove()` will remove the service from the registry and is only called in the case of a watcher panic, any subsequent call will add the service again and start a watcher. `expire()` will clear all nodes/entries from the cache for the given service, this is invoked when calls to the agent are erroring but the watcher is still active and will continue attempting calls to the agent on a backoff retry

File: nameresolution/consul/consul.go
Comment (artursouza) [reply nameresolution/consul/consul.go:106]: To shuffle, sure. In this case, it is to pick one - which a simple random number with mod operation should accomplish the same.

Comment (a-elsheikh) [reply nameresolution/consul/consul.go]: I've made the change - is that what you had in mind?

**Comment by a-elsheikh** [comment]
@artursouza can you have a look please

**Comment by artursouza** [comment]
Reopening

**Comment by chenjpu** [comment]
image

**Comment by a-elsheikh** [comment]
> image

@yaron2 @artursouza any updates please :)

**Comment by badgeratu** [comment]
@artursouza @yaron2 I believe we addressed your review comments on this PR some time ago, with 1.8.0 now out the door could you please pick this up and prioritize it again as we do need this for our next phase of Dapr rollout.

**Comment by berndverst** [comment]
@a-elsheikh could you please merge master into the branch again to address the potential conflict? I tried pushing a commit but it does not appear you are giving maintainers edit access to your PR / branch.

**Comment by badgeratu** [comment]
It's definitely not outdated, we've been waiting patiently for a re-review as we believe we've addressed all comments. If you could push that it would be greatly appreciated, this is still a vital feature required by us.

**Comment by ItalyPaleAle** [comment]
Ok, thanks for confirming. If you can fix the merge conflicts, I'll review it for Dapr 1.11.

**Comment by alberthuang24** [comment]
I also really need this feature

**Comment by berndverst** [comment]
@badgeratu because we do not have write access to your fork we cannot shepherd this PR -- so this PR will constantly encounter merge conflicts that only you can resolve.
It's very difficult to merge PRs where we as maintainers do not have automatic write access to the remote fork.

Can you invite contrib maintainers directly to have write access to your fork so we can unblock this pr? The maintainers are: @ItalyPaleAle @yaron2 @berndverst @artursouza 

We are days (or maybe just one day) away from code freeze for 1.11. If we find that tests pass after addressing merge conflicts and updating the branch then this is likely good to go - but we'll re-review then.

**Comment by a-elsheikh** [comment]
hey @berndverst thanks for reaching out - as this fork exists under the corporate area we've limited flexibility in amending the permissions. In any case, I've got eyes on this now so am happy to iterate with the maintainers to see it done now that there is engagement.

**Comment by berndverst** [comment]
@a-elsheikh @badgeratu can you please ping me in Discord (`@bernd`) once you have resolved merge conflicts and fixed DCO here?

I will try to get this merged - we are now at code freeze for 1.12 - but if you can get this resolved Monday we can still try to get it in!

**Comment by ItalyPaleAle** [comment]
Closing as it's been re-created as #3121

**Pull request closed without merging**
